service/history/stateBuilder.go

```diff
@@ -658,12 +658,22 @@ func (b *stateBuilderImpl) scheduleWorkflowTimerTask(event *shared.HistoryEvent,
 	timeout := now.Add(time.Duration(msBuilder.GetExecutionInfo().WorkflowTimeout) * time.Second)
 
 	cronSchedule := b.msBuilder.GetExecutionInfo().CronSchedule
-	cronBackoffDuration := backoff.GetBackoffForNextSchedule(cronSchedule, now)
-	if cronBackoffDuration != backoff.NoBackoff {
-		timeout = timeout.Add(cronBackoffDuration)
+	backoffDuration := backoff.NoBackoff
+	startEvent, ok := msBuilder.GetStartEvent()
+	if ok &&
+		startEvent.WorkflowExecutionStartedEventAttributes != nil &&
+		startEvent.WorkflowExecutionStartedEventAttributes.FirstDecisionTaskBackoffSeconds != nil {
+		backoffDuration = time.Duration(*startEvent.WorkflowExecutionStartedEventAttributes.FirstDecisionTaskBackoffSeconds) * time.Second
+	}
+	if backoffDuration != backoff.NoBackoff {
+		timeout = timeout.Add(backoffDuration)
+		timeoutType := persistence.WorkflowBackoffTimeoutTypeRetry
+		if len(cronSchedule) > 0 {
+			timeoutType = persistence.WorkflowBackoffTimeoutTypeCron
```

Comment:
This is kind of interesting... I don't kind you can determine the type just by looking cronSchedule. For example, a cron workflow with retry policy and needs retry.

Actually, you fixed another bug. Seems like previously we won't create this backoff timer at passive side if workflow needs retry. Maybe do some local verification or add a xdc integration test?

Reply:
done